test/unit/carousel.ts

```diff
@@ -0,0 +1,414 @@
+import { utils } from '@storefront/core';
+import * as sinon from 'sinon';
+import Carousel from '../../src/carousel';
+import suite from './_suite';
+
+suite('Carousel', ({ expect, spy, stub }) => {
+
+  describe('settings', () => {
+    it('should default to default settings if no settings are passed in', () => {
+      const carousel = new Carousel();
+
+      expect(carousel.props.speed).to.eq(0);
+      expect(carousel.props.slidestoscroll).to.eq(1);
+      expect(carousel.props.slidestoshow).to.eq(1);
+    });
+  });
+
+  describe('functions', () => {
+    let carousel: Carousel;
+    beforeEach(() => {
+      carousel = new Carousel();
+      carousel.props.items = [
+        { content: 'test0' },
+        { content: 'test1' },
+        { content: 'test2' },
+        { content: 'test3' },
+        { content: 'test4' },
+        { content: 'test5' },
+        { content: 'test6' },
+        { content: 'test7' },
+        { content: 'test8' },
+        { content: 'test9' },
+      ];
+    });
+
+    describe('onMount()', () => {
+      it('should add event listener for window resize', () => {
+        const addEventListener = spy();
+        stub(utils, 'WINDOW').returns({ addEventListener });
+
+        carousel.onMount();
+
+        expect(addEventListener).to.be.calledWithExactly('resize', carousel.update);
+      });
+    });
+
+    describe('onUpdate()', () => {
+      it('should add cloned items to state', () => {
+        const clone = carousel.cloneItems = spy();
+
+        carousel.onUpdate();
+
+        expect(clone).to.be.called;
+      });
+    });
+
+    describe('onUnmount()', () => {
+      it('should add event listener for window resize', () => {
+        const removeEventListener = spy();
+        stub(utils, 'WINDOW').returns({ removeEventListener });
+
+        carousel.onUnmount();
+
+        expect(removeEventListener).to.be.calledWithExactly('resize', carousel.update);
+      });
+    });
+
+    describe('moveNext', () => {
+      it('should call slideHandler function with correct slide number', () => {
+        carousel.state.currentSlide = 0;
+        carousel.slideHandler = spy();
+
+        carousel.moveNext();
+
+        expect(carousel.slideHandler).to.be.calledWithExactly(1);
```

Comment:
I don't like to do expectations around objects that the tests don't own if it can be avoided.
This isn't a problem in this specific instance, but the pattern we use ensures we will never accidentally have this problem.

I would want
```js
const slideHandler = carousel.slideHandler = spy();

// ...

expect(slideHandler).to.be.calledWithExactly(1);
```

This avoids complications where as a result of your test's action (in this case, `carousel.moveNext()`) possibly manipulates or re-assigns `carousel.slideHandler` to a new value and you end up doing expectations against the wrong value

Please make this change to all of your tests in order to increase their resilience
Come to me if you need more context on why we're using this pattern.